src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,382 @@
+                    //    + " init: " + init.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " cutShards: " + cutShards.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " kv.of: " + kving.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " output: " + outputting.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " rejected " + rejected);
+                }
+
+                private boolean accept(GATKRead r, SimpleInterval region) {
+                    if (r.isUnmapped()) {
+                        // we don't want unmapped reads, *unless* their mate is in the region
+                        return (!r.mateIsUnmapped() && r.getMateStart() >= region.getStart());
+                    } else {
+                        // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                        return (r.getStart() >= region.getStart());
+                    }
+                }
+
+                private boolean accept(SAMRecord r, SimpleInterval region) {
+                    if (r.getReadUnmappedFlag() ||
+                        r.getReferenceName() == null || r.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME) ||
+                        r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+                        // unmapped read. We don't want it, *unless* their mate is in the region
+                        return (!r.getMateUnmappedFlag() && r.getMateAlignmentStart() >= region.getStart());
+                    } else {
+                        // mapped read.
+                        // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                        return (r.getStart() >= region.getStart());
+                    }
+                }
+
+
+            }));
+    }
+
+
+    // chr2:1-200 -> chr2:1-100,chr2:101-200
+    static public List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize) {
+        ArrayList<SimpleInterval> ret = new ArrayList<SimpleInterval>();
+        for (SimpleInterval i : intervals) {
+            int beginShard = shardIndex(i.getStart(), shardSize);
+            int endShard = shardIndex(i.getEnd(), shardSize);
+            if (beginShard==endShard) {
+                ret.add(i);
+                continue;
+            }
+            // more than one shard: output begin to end-of-shard, then multiple full shards, then begin-of-shard to end.
+            ret.add(new SimpleInterval(i.getContig(), i.getStart(), endOfShard(beginShard, shardSize)));
+            for (int shard = beginShard+1; shard<endShard; shard++) {
+                ret.add(new SimpleInterval(i.getContig(), beginOfShard(shard, shardSize), endOfShard(shard, shardSize)));
+            }
+            ret.add(new SimpleInterval(i.getContig(), beginOfShard(endShard, shardSize), i.getEnd()));
+        }
+        return ret;
+    }
```

Comment:
This `cutToShards()` methods strikes me as exactly the kind of method that would have edge-case bugs -- could you add good unit tests for this method in isolation?